app/com/arpnetworking/metrics/portal/reports/ReportExecution.java

```diff
@@ -0,0 +1,184 @@
+            final Injector injector,
+            final S source,
+            final F format,
+            final Instant scheduled
+    ) {
+        return getRenderer(injector, source, format).render(source, format, scheduled);
+    }
+
+    /* package private */ static CompletionStage<Void> sendAll(
+            final Injector injector,
+            final ImmutableMultimap<Recipient, ReportFormat> recipientToFormats,
+            final ImmutableMap<ReportFormat, RenderedReport> formatToRendered
+    ) {
+        final CompletableFuture<?>[] futures = recipientToFormats
+                .asMap()
+                .entrySet()
+                .stream()
+                .map(entry -> getSender(injector, entry.getKey())
+                        .send(entry.getKey(), mask(formatToRendered, entry.getValue()))
+                        .toCompletableFuture())
+                .toArray(CompletableFuture[]::new);
+        return CompletableFuture.allOf(futures);
+    }
+
+    private static <S extends ReportSource, F extends ReportFormat> Renderer<S, F> getRenderer(
+            final Injector injector,
+            final S source,
+            final F format
+    ) {
+        final String keyName = getRendererKeyName(source, format);
+        @SuppressWarnings("unchecked")
+        final Renderer<S, F> renderer = injector.getInstance(Key.get(Renderer.class, Names.named(keyName)));
+        if (renderer == null) {
+            throw new IllegalArgumentException("no Renderer exists for key name '" + keyName + "'");
+        }
+        return renderer;
+    }
+
+    private static Sender getSender(final Injector injector, final Recipient recipient) {
+        final String keyName = getSenderKeyName(recipient);
+        final Sender sender = injector.getInstance(Key.get(Sender.class, Names.named(keyName)));
+        if (sender == null) {
+            throw new IllegalArgumentException("no Sender exists for key name '" + keyName + "'");
+        }
+        return sender;
+    }
+
+    /* package private */ static String getSenderKeyName(final Recipient recipient) {
+        return CaseFormat.UPPER_UNDERSCORE.to(CaseFormat.LOWER_UNDERSCORE, recipient.getType().name());
+    }
+
+    /* package private */ static String getRendererKeyName(final ReportSource source, final ReportFormat format) {
+        return SOURCE_TYPE_VISITOR.visit(source) + " " + FORMAT_TYPE_VISITOR.visit(format);
+    }
+
+    /**
+     * Restrict {@code map} to key/value pairs whose keys are in {@code keys}.
+     */
+    private static <K, V> ImmutableMap<K, V> mask(final ImmutableMap<K, V> map, final Collection<K> keys) {
+        return keys.stream().collect(ImmutableMap.toImmutableMap(key -> key, map::get));
+    }
+
+    private ReportExecution() {}
+
+    private static final ReportSource.Visitor<String> SOURCE_TYPE_VISITOR = new ReportSource.Visitor<String>() {
+        @Override
+        public String visit(final ChromeScreenshotReportSource source) {
```

Comment:
We're going to be renaming this source right?

Reply:
Yep! I was hoping to put that in another diff, since it's far-reaching.

Reply:
No problem. Yes, no one is using this yet.